pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,794 @@
+	}
+
+	return setting, nil
+}
+
+// webhook needs check if VMs are off
+func (h *Handler) OnStorageNetworkChange(key string, setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
+	if setting == nil || setting.DeletionTimestamp != nil || setting.Name != settings.StorageNetworkName {
+		return nil, nil
+	}
+	settingCopy := setting.DeepCopy()
+
+	if settingCopy.Annotations == nil {
+		settingCopy.Annotations = make(map[string]string)
+	}
+
+	var (
+		updatedSetting *harvesterv1.Setting
+		err            error
+		value          string
+	)
+
+	if updatedSetting, err = h.checkValueIsChanged(settingCopy); err != nil {
+		return updatedSetting, err
+	}
+
+	value, err = h.getLonghornStorageNetwork()
+	if err != nil {
+		return setting, err
+	}
+
+	currentNad := setting.Annotations[NadStorageNetworkAnnotation]
+	if currentNad == value {
+		// if post config is successful, it will finish the onChange
+		return h.handleLonghornSettingPostConfig(settingCopy)
+	}
+
+	logrus.Infof("storage network change: %s", settingCopy.Value)
+
+	// if replica eq 0, skip
+	// save replica to annotation
+	// set replica to 0
+	if err = h.checkPodStatusAndStop(); err != nil {
+		logrus.Infof("Requeue to check pod status again")
+		updatedSetting, updateConditionErr := h.setConfiguredCondition(settingCopy, false, ReasonInProgress, MsgStopPod)
+		if updateConditionErr != nil {
+			return setting, fmt.Errorf("update status error %v", updateConditionErr)
+		}
+		return updatedSetting, err
+	}
+
+	// check volume detach before put LH settings
+	if err = h.checkLonghornVolumeDetached(); err != nil {
+		updatedSetting, updateConditionErr := h.setConfiguredCondition(settingCopy, false, ReasonInProgress, err.Error())
+		if updateConditionErr != nil {
+			return setting, fmt.Errorf("update status error %v", updateConditionErr)
+		}
+		return updatedSetting, err
+	}
+
+	logrus.Infof("all pods are stopped")
+	logrus.Infof("all volumes are detached")
+	logrus.Infof("update Longhorn settings")
+	// push LH setting
+	nadName := settingCopy.Annotations[NadStorageNetworkAnnotation]
+	if err = h.updateLonghornStorageNetwork(nadName); err != nil {
+		return setting, fmt.Errorf("Update Longhorn setting error %v", err)
+	}
+	if updatedSetting, err = h.setConfiguredCondition(settingCopy, false, ReasonInProgress, MsgUpdateLonghornSetting); err != nil {
+		return setting, fmt.Errorf("update status error %v", err)
+	}
+
+	return updatedSetting, nil
+}
+
+// calc sha1 hash
+func (h *Handler) sha1(s string) string {
+	hash := sha1.New()
+	hash.Write([]byte(s))
+	sha1sum := hash.Sum(nil)
+	return fmt.Sprintf("%x", sha1sum)
+}
+
+func (h *Handler) checkIsSameHashValue(setting *harvesterv1.Setting) bool {
+	currentHash := h.sha1(setting.Value)
+	savedHash := setting.Annotations[HashStorageNetworkAnnotation]
+	return currentHash == savedHash
+}
+
+func (h *Handler) setHashAnnotations(setting *harvesterv1.Setting) {
+	setting.Annotations[HashStorageNetworkAnnotation] = h.sha1(setting.Value)
```

Comment:
This will make another case weird.
Another case will also call setHashAnnotaions but it still needs to save and replace the NAD annotation by itself but not via this method.

Reply:
I will consider to move those NAD annotations to another function rather than putting it in setHashAnnotaions.
setHashAnnotations did the different things with NAD

Reply:
fix with another method